### DISCUSSION TITLE
xit! for Neovim

### DISCUSSION THREAD

**Discussion by synaptiko** [jotaen/xit#26] [Show & tell]
Hello guys,

I'm also jumping into this [x]it! wave, I suppose :-)

I started by implementing (very) basic tree-sitter grammar: https://github.com/synaptiko/tree-sitter-xit

I plan to start using it in https://github.com/synaptiko/xit.nvim for additional things like highlighting, quick jumps between tasks (in a way that the cursor would always be at the start of task name, for example), extraction of all finished/unfinished tasks (useful for reports).

It's my very first time implementing with tree-sitter & lua in Neovim, it's definitely not in a good quality, so naturally any PRs are welcomed.

_Note: Neovim 0.7 is required (it should be released officially in 1-2 weeks, AFAIK)._

**Comment by jotaen** [comment]
👍 Awesome project, a proper tree-sitter grammar will be great for re-usability in other editors and tools.

I’m happy to link both projects from [the website](https://xit.jotaen.net/). It’s just an offer, so don’t feel pushed. It might help other users to discover tools more easily, though. Let me know if (or when) I should add them.

I don’t know whether you have seen it, so just in case: I’ve recently published a comprehensive [syntax guide](https://xit.jotaen.net/syntax-guide). It’s not 100% complete yet, but it should already cover a lot of corner cases of the syntax.

**Reply by synaptiko** [reply]
Thank you and thanks for the offer. I think it needs a bit more time before publishing "officially" as there are still plenty of missing parts.

I actually used your guide, it was really helpful. Although I have to admit, some parts are a bit tricky, especially everything related to whitespaces. I'll try to cover it as closely as possible :slightly_smiling_face: I'm covering it with tests like these: https://github.com/synaptiko/tree-sitter-xit/blob/master/test/corpus/simple-tasks.txt

**Reply by jotaen** [reply]
If you have any feedback or thoughts on any of these whitespace things, let me know. The spec should be easy to work with for both users and tool creators, so anything that makes that process easier is welcomed.

Also, some of the details, e.g. how tags or due dates can appear in descriptions, still need more clarification. See [here](https://github.com/jotaen/xit/issues/28) or [here](https://github.com/jotaen/xit/issues/27), for example.

Other than that, I think it won’t be possible (or at least pretty hard) to get *all* rules correct with a syntax highlighting engine or grammar. E.g. if you consider that “only the first due date in a description should be recognised”, or “due dates must be representable by the gregorian calendar”. But I think that’s alright.

**Reply by synaptiko** [reply]
I've rewritten the rules a bit and whitespace/newline problems are resolved now. Resulting in the following:
![image](https://user-images.githubusercontent.com/1593280/162582083-1d5aceb8-6826-4469-98cc-50139ecc30ab.png)

But what I'm currently fighting with is priorities; I've created [a branch with the basic test](https://github.com/synaptiko/tree-sitter-xit/tree/priority) but I'm not able to figure out [a combination of rules](https://github.com/synaptiko/tree-sitter-xit/blob/priority/grammar.js) to make it work. The complication is in the fact that the task name can also contain `!`, `.` and spaces. I tried usage of `conflicts`, `prec.*` etc. but without luck. What partially worked was when I said that `main_line` can't start with space or `!` but then "Absent description" test case started failing. I assume I'll need to learn a bit more about tree-sitter to crack this nut :-)

_Note: I've decided to try my luck on Stackoverflow: https://stackoverflow.com/questions/71810032/tree-sitter-problem-with-optional-tokens-which-has-special-meaning-but-their-ch_

**Reply by synaptiko** [reply]
One more thing I found hard to implement with tree-sitter (most probably because of my lack of knowledge in this area) are groups. I almost feel that some of these things can't be modeled with what tree-sitter offers. I think my initial version will use some more relax rules, ie. I was able to recognize headline, but without groups yet. This should be usable for highlighting, at least:
![image](https://user-images.githubusercontent.com/1593280/162585181-2b987fda-6b4a-42ac-87f8-9727d82e34eb.png)

**Reply by synaptiko** [reply]
I think it will be necessary to use https://tree-sitter.github.io/tree-sitter/creating-parsers#external-scanners in order to handle some of special cases, like groups. It looks like that the scanner has access to next character with `lookahead` argument. Unfortunately scanner can only be written in C/C++ which is a bummer for me as I'm not good in these two :disappointed:

**Reply by synaptiko** [reply]
I was able to finally solve the priorities in a reasonable way :tada:
![image](https://user-images.githubusercontent.com/1593280/162590310-2c846017-af6d-4dba-8aeb-d112f12ebc89.png)

**Comment by synaptiko** [comment]
FYI I published a programming session video where I refactored the grammar a bit, I also tried to explaine the plugin itself & the tree-sitter (at least my understanding of some of the things):
https://www.youtube.com/watch?v=sYGXiTi63hI&list=PLJU9kDne-X6pAMC1NOeNnOfLjnVM25dot

If I'll have more time, I'll record more and publish it in the list :-)

**Reply by jotaen** [reply]
Cool, thanks for sharing this. I have never worked with tree sitter either, so I’m curious to check this out.

**Comment by synaptiko** [comment]
@jotaen I'm happy to tello you that I just released `v0.1` of the plugin: https://github.com/synaptiko/xit.nvim

The features can be seen here in action: https://www.youtube.com/watch?v=VrFdujDqeOA

So if you want, you can add Neovim to the list of supported editors. I think this is now pretty much usable :slightly_smiling_face:

**Reply by jotaen** [reply]
Very cool, I’ve added it to the list.

I also love how your roadmap is written in xit, that’s very meta 😎

**Comment by hern42** [comment]
hi, 
i'm sorry i'm new to neovim and the package management and installation...
i used vim.plug to install xit and tree.sitter.
however i'm pretty sure i forgot to write something in my .config/nvim/init.vim because nothing works and as i'm not using packer i cannot just copy/paste the config.
i'd be happy if someone could point me in the good direction... something to add after the accolades in "Plug 'synaptiko/xit.nvim', {   }"
all the very best,
a-

**Reply by hern42** [reply]
here is the part in the config file 

 23 call plug#begin('~/.config/nvim/plugged')
 24 │" Plugin Section
 25 │Plug 'scrooloose/nerdtree'                             
 26 │Plug 'lukas-reineke/indent-blankline.nvim'
 27 │Plug 'synaptiko/xit.nvim' 
 28 │Plug 'nvim-treesitter/nvim-treesitter', {'do': ':TSUpdate'}
 29 call plug#end()
 30 
 31 " Start NERDTree when Vim is opened and put the cursor back in the other window.
 32 autocmd VimEnter * NERDTree | wincmd p

**Reply by jotaen** [reply]
Not sure here is the right place to ask, since this repo only contains the file specification of [x]it! itself. You could try to open an issue in the https://github.com/synaptiko/xit.nvim repository directly.

**Reply by hern42** [reply]
i will, thanks...
i'm sorry, now that you mention it it is indeed the wrong place to ask.
best...
